llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -1306,6 +1306,10 @@ void PassBuilder::addVectorPasses(OptimizationLevel Level,
   }
   // Cleanup after the loop optimization passes.
   FPM.addPass(InstCombinePass());
+  // InstCombine can create CSE opportunities when it cleans the result of loop
+  // vectorization. They occur when combines use replaceOperand, which happens
+  // most often when combining the boolean operations created by if-conversion.
+  FPM.addPass(EarlyCSEPass());
 
   if (Level.getSpeedupLevel() > 1 && ExtraVectorizerPasses) {
     ExtraFunctionPassManager<ShouldRunExtraVectorPasses> ExtraPasses;
```

Comment:
```suggestion
  // InstCombine can create CSE opportunities when it cleans the result of loop
  // vectorization. They occur when combines use replaceOperand, which happens
  // most often when combining the boolean operations created by if-conversion.
  ExtraFunctionPassManager<ShouldRunExtraVectorPasses> ExtraPasses;
  ExtraPasses.addPass(EarlyCSEPass());

  if (Level.getSpeedupLevel() > 1 && ExtraVectorizerPasses) {
    ExtraFunctionPassManager<ShouldRunExtraVectorPasses> ExtraPasses;
    ....
  }
 FPM.addPass(std::move(ExtraPasses));
```

Would probably help to soften the compile-time blow a bit

Reply:
Thanks @fhahn.  For my education is `ShouldRunExtraVectorPasses` something LoopVectorize "sets" to highlight the potential need for extra optimisations?